src/commands/audit.ts

```diff
@@ -0,0 +1,112 @@
+import { WorkspaceRequiredError } from '@yarnpkg/cli'
+import { CommandContext, Configuration, Project, treeUtils } from '@yarnpkg/core'
+import { Command, Usage, Option, UsageError } from 'clipanion'
+import { getTree, focusWorkspaces } from '../utils'
+
+export class LicensesAuditCommand extends Command<CommandContext> {
+  static paths = [[`licenses`, `audit`]]
+
+  recursive = Option.Boolean(`-R,--recursive`, false, {
+    description: `Include transitive dependencies (dependencies of direct dependencies)`
+  })
+
+  production = Option.Boolean(`--production`, false, {
+    description: `Exclude development dependencies`
+  })
+
+  json = Option.Boolean(`--json`, false, {
+    description: `Format output as JSON`
+  })
+
+  excludeMetadata = Option.Boolean(`--exclude-metadata`, false, {
+    description: `Exclude dependency metadata from output`
+  })
+
+  focus = Option.Array(`--focus`, [], {
+    description: `Focus on one or more workspaces`
+  })
+
+  allowedLicenses = Option.Array(`--allowed`, [], {
+    description: `List of allowed licenses`
+  })
+
+  blockedLicenses = Option.Array(`--blocked`, [], {
+    description: `List of blocked licenses`
+  })
+
+  exitCode = Option.Boolean(`--exit-code`, false, {
```

Comment:
what's the use case for this new flag? i think we should always exit non-zero if there are violations. if the user doesn't like that, they can ignore the non-zero exit code themselves.

Reply:
I found testing this to be very annoying because when the exit code is non-zero:
- jest API to check for errors is very limited, it's not possible to use `toThrowErrorMatchingSnapshot` because the error thrown is stringified to `"Command failed: yarn licenses audit ..."`
- if using `toThrowError(expect.objectContaining({ ... }))` we can't use snapshots
- for reasons I don't know, `spawnSync` (or the caught error from `execSync`) will _always_ return `null` for `status` and `stdout`
So far I didn't find a way that reliably returns the stdout and the status in a way that allows me to snapshot test the stdout.

Reply:
It seems that passing `shell: true` to `spawnSync` fixes this. The documentation on this option is a bit lacking, but it should be safe to use only for testing.